rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/WorkerBookKeeper.java

```diff
@@ -52,4 +132,41 @@ private void startHeartbeatService(Configuration conf, MetricRegistry metrics, B
     this.heartbeatService = new HeartbeatService(conf, metrics, factory, this);
     heartbeatService.startAsync();
   }
+
+  /**
+   * Attempt to initialize the client for communicating with the master BookKeeper.
+   *
+   * @param bookKeeperFactory   The factory to use for creating a BookKeeper client.
+   * @return The client used for communication with the master node.
+   */
+  private static RetryingBookkeeperClient initializeClientWithRetry(BookKeeperFactory bookKeeperFactory, Configuration conf,
+                                                             String hostName)
+  {
+    final int retryInterval = CacheConfig.getServiceRetryInterval(conf);
+    final int maxRetries = CacheConfig.getServiceMaxRetries(conf);
+
+    for (int failedStarts = 0; failedStarts < maxRetries; ) {
```

Comment:
why do you need to do all the retries here? RetryingBookkeeperClient is meant to do it.

Reply:
The retries are needed because even with the retries in RetryingBookkeeperClient, the worker might not be able to connect to master. Reason being master bookkeeper might not be up whereas the worker bookkeeper is up. This call is only made during the start of the worker bookkeeper.

Reply:
In that case we need to push the transport creation and opening into RetryingBookkeeperClient from BookKeeperFactory.createBookKeeperClient and extend RetryingBookkeeperClient retry logic to also take sleep interval.

Reply:
Currently the retry logic in RetryingBookkeeperClient is only during the method invoke. We create the transport and open it without any retry. So createBookKeeperClient doesn't have any retry logic.
I agree with you that all the retry logic should be in one place and that retry should take sleep interval to make it work in all the cases. But I think that is outside the scope of this PR. So in my opinion, we should open a new issue to address this problem.

Reply:
Updated 

> In that case we need to push the transport creation and opening into RetryingBookkeeperClient from BookKeeperFactory.createBookKeeperClient and extend RetryingBookkeeperClient retry logic to also take sleep interval.

Done

Reply:
@jordanw-bq I have changed the retrying logic of creating the bookkeeper client. Let me know if you have any comments